File: app/src/main/java/com/earthgame/idle/domain/challenges/Challenges.kt
package com.earthgame.idle.domain.challenges

import com.earthgame.idle.domain.engine.gd
import com.earthgame.idle.domain.model.GameState
import com.earthgame.idle.domain.model.GasId
import com.earthgame.idle.domain.model.ResourceId
import com.earthgame.idle.domain.prestige.PrestigeUpgradeEffect
import com.earthgame.idle.domain.technologies.ALL_TECHNOLOGIES
import com.earthgame.idle.domain.technologies.TECH_BY_ID
import com.earthgame.idle.domain.technologies.TechBranch
import com.earthgame.idle.domain.technologies.Technology

data class ChallengeRestriction(
    /** Technologies that cannot be purchased while this challenge is active. */
    val disabledTechIds: Set<String> = emptySet(),
    /** Technologies may not exceed this tier (e.g. "Primitive": stop at the Iron Age). */
    val maxTechTier: Int? = null,
    /** The run fails if temperature ever exceeds this. */
    val maxTemperatureC: Double? = null,
    /** Only technologies producing exclusively these gases (plus gasless unlocks) may be purchased. */
    val onlyGasIds: Set<GasId>? = null,
)

data class ChallengeGoal(
    val description: String,
    val check: (state: GameState, civLevel: Int) -> Boolean,
)

data class Challenge(
    val id: String,
    val displayName: String,
    val description: String,
    val icon: String,
    val restriction: ChallengeRestriction,
    val goal: ChallengeGoal,
    val rewardDescription: String,
    val reward: PrestigeUpgradeEffect,
)

private fun techProducesOnly(tech: Technology, allowedGases: Set<GasId>): Boolean =
    tech.effect.gasProductionPerUnit.keys.all { it in allowedGases }

val CHALLENGES: List<Challenge> = listOf(
    Challenge(
        id = "ice_age",
        displayName = "Ice Age",
        description = "Reach a strong industrial base while keeping temperature below +2°C.",
        icon = "🧊",
        restriction = ChallengeRestriction(maxTemperatureC = 2.0),
        goal = ChallengeGoal("Reach civilization level 40 without exceeding +2°C.") { _, civLevel -> civLevel >= 40 },
        rewardDescription = "+10% permanent production",
        reward = PrestigeUpgradeEffect(globalProductionMultiplier = 1.1),
    ),
    Challenge(
        id = "low_carbon",
        displayName = "Low Carbon",
        description = "Coal technologies are forbidden for the entire run.",
        icon = "🚫",
        restriction = ChallengeRestriction(
            disabledTechIds = ALL_TECHNOLOGIES.filter { it.id.contains("coal") }.map { it.id }.toSet(),
        ),
        goal = ChallengeGoal("Accumulate 1 trillion (1e12) total Energy produced.") { state, _ ->
            state.resources[ResourceId.ENERGY].gte(gd("1e12"))
        },
        rewardDescription = "+25% Research, permanently",
        reward = PrestigeUpgradeEffect(researchMultiplier = 1.25),
    ),
    Challenge(
        id = "no_oil",
        displayName = "No Oil",
        description = "Oil and its downstream technologies are off-limits.",
        icon = "🛢️",
        restriction = ChallengeRestriction(
            disabledTechIds = ALL_TECHNOLOGIES.filter {
                it.branch == TechBranch.FOSSIL_FUELS ||
                    it.id in setOf("automobile", "diesel_engine", "jet_aircraft", "petrochemicals")
            }.map { it.id }.toSet(),
        ),
        goal = ChallengeGoal("Reach civilization level 35 without any oil technology.") { _, civLevel -> civLevel >= 35 },
        // The reference implementation's copy here advertised a Transportation
        // branch bonus that its own effect never granted. The effect is the
        // shipped balance and is left exactly as it is; the text is what was
        // wrong, so the text is what changed.
        rewardDescription = "+5% permanent production",
        reward = PrestigeUpgradeEffect(globalProductionMultiplier = 1.05),
    ),
    Challenge(
        id = "primitive",
        displayName = "Primitive",
        description = "Never advance past the Iron Age.",
        icon = "⛏️",
        restriction = ChallengeRestriction(maxTechTier = 6),
        goal = ChallengeGoal("Reach 400 ppm atmospheric CO₂ using only Primitive-era technology.") { state, _ ->
            state.atmosphere[GasId.CO2].gte(gd(120))
        },
        rewardDescription = "Start every future run with extra starting Energy",
        reward = PrestigeUpgradeEffect(startingResources = mapOf(ResourceId.ENERGY to 500.0)),
    ),
    Challenge(
        id = "speedrun",
        displayName = "Speedrun",
        description = "Reset Earth as quickly as possible.",
        icon = "⏱️",
        restriction = ChallengeRestriction(),
        // Scored against the run's own elapsed time rather than a wall clock
        // read inside the check, so the same state always gives the same answer.
        goal = ChallengeGoal("Reset within 15 minutes of starting the run.") { state, _ ->
            (state.lastTickAt - state.runStartedAt) / 1000.0 <= 900
        },
        // A ×1.0 multiplier is not a bonus, and the reference implementation's
        // copy promised one anyway: there is no "Earth Points earned" field in
        // PrestigeUpgradeEffect for it to have been wired to. Rather than
        // invent a balance change, the reward is described for what it is —
        // adding that field is tracked in docs/CODEBASE_AUDIT.md.
        rewardDescription = "Bragging rights — this one pays no permanent bonus",
        reward = PrestigeUpgradeEffect(globalProductionMultiplier = 1.0),
    ),
    Challenge(
        id = "single_gas",
        displayName = "Single Gas",
        description = "Only technologies that emit CO₂ (or nothing) may be used.",
        icon = "☁️",
        restriction = ChallengeRestriction(
            disabledTechIds = ALL_TECHNOLOGIES
                .filterNot { techProducesOnly(it, setOf(GasId.CO2)) }
                .map { it.id }.toSet(),
        ),
        goal = ChallengeGoal("Reach civilization level 30 using only CO₂-producing technology.") { _, civLevel -> civLevel >= 30 },
        rewardDescription = "+30% CO₂ production, permanently",
        reward = PrestigeUpgradeEffect(gasProductionMultiplier = GasId.CO2 to 1.3),
    ),
    Challenge(
        id = "methane_world_challenge",
        displayName = "Methane World",
        description = "Make CH₄ the dominant greenhouse gas by the time you reset.",
        icon = "🐄",
        restriction = ChallengeRestriction(),
        goal = ChallengeGoal("Reset while CH₄ contributes more forcing than any other gas.") { state, _ ->
            state.forcing.perGas[GasId.CH4] > state.forcing.perGas[GasId.CO2] &&
                state.forcing.perGas[GasId.CH4] > 0
        },
        rewardDescription = "+50% CH₄ production, permanently",
        reward = PrestigeUpgradeEffect(gasProductionMultiplier = GasId.CH4 to 1.5),
    ),
    Challenge(
        id = "zero_emissions",
        displayName = "Zero Emissions",
        description = "Build a strong civilization while keeping gross emissions minimal.",
        icon = "🌱",
        restriction = ChallengeRestriction(
            disabledTechIds = setOf("coal_power_plant", "supercritical_coal", "coal_mega_mining", "oil_sands"),
        ),
        // The goal text is the reference implementation's and is pinned by
        // `ContentParityTest`; the emission ceiling it mentions is enforced by
        // the disabled-technology list above rather than by a live check on the
        // production rate. See docs/CODEBASE_AUDIT.md.
        goal = ChallengeGoal("Reach civilization level 25 while total gross gas production stays under 1e6 kg/s.") { _, civLevel ->
            civLevel >= 25
        },
        rewardDescription = "+20% Research, permanently",
        reward = PrestigeUpgradeEffect(researchMultiplier = 1.2),
    ),
)

val CHALLENGE_BY_ID: Map<String, Challenge> = CHALLENGES.associateBy { it.id }

/** Whether purchasing [techId] would violate [challenge]'s restriction. */
fun isTechRestrictedByChallenge(challenge: Challenge, techId: String): Boolean {
    val tech = TECH_BY_ID[techId] ?: return false
    if (techId in challenge.restriction.disabledTechIds) return true
    challenge.restriction.maxTechTier?.let { if (tech.tier > it) return true }
    val onlyGases = challenge.restriction.onlyGasIds
    if (onlyGases != null && tech.effect.gasProductionPerUnit.isNotEmpty()) {
        if (!techProducesOnly(tech, onlyGases)) return true
    }
    return false
}

/**
 * The full set of technologies an active challenge locks out, precomputed once
 * per state change rather than per card render.
 */
fun disabledTechIdsForChallenge(challenge: Challenge?): Set<String> {
    if (challenge == null) return emptySet()
    return ALL_TECHNOLOGIES.filter { isTechRestrictedByChallenge(challenge, it.id) }.map { it.id }.toSet()
}

/** Whether the challenge's live restriction (e.g. a temperature ceiling) has been broken this run. */
fun isChallengeRestrictionViolated(challenge: Challenge, state: GameState): Boolean {
    val ceiling = challenge.restriction.maxTemperatureC ?: return false
    return state.temperatureAnomalyC > ceiling
}

fun isChallengeGoalMet(challenge: Challenge, state: GameState, civLevel: Int): Boolean =
    challenge.goal.check(state, civLevel)

/** Every completed challenge's permanent reward, in the shape prestige upgrades use. */
fun computeChallengeRewardEffects(completed: Map<String, Boolean>): List<PrestigeUpgradeEffect> =
    CHALLENGES.filter { completed[it.id] == true }.map { it.reward }

File: tools/ts-reference/engine/challenges.ts
import { GameState } from './gameState';
import { ALL_TECHNOLOGIES, TECH_BY_ID, Technology } from './technologies';
import { GasId } from './gases';
import { PrestigeUpgradeEffect } from './prestige';

export interface ChallengeRestriction {
  /** Technologies that cannot be purchased while this challenge is active. */
  disabledTechIds?: string[];
  /** Technologies may not exceed this tier (e.g. "Primitive": stop at the Iron Age). */
  maxTechTier?: number;
  /** The run fails (restriction violated) if temperature ever exceeds this. */
  maxTemperatureC?: number;
  /** Only technologies producing exclusively these gases (plus gasless unlocks) may be purchased. */
  onlyGasIds?: GasId[];
}

export interface ChallengeGoal {
  description: string;
  check: (state: GameState, civLevel: number) => boolean;
}

export interface Challenge {
  id: string;
  name: string;
  description: string;
  icon: string;
  restriction: ChallengeRestriction;
  goal: ChallengeGoal;
  rewardDescription: string;
  reward: PrestigeUpgradeEffect;
}

function techProducesOnly(tech: Technology, allowedGases: GasId[]): boolean {
  const produced = Object.keys(tech.effect.gasProductionPerUnit ?? {}) as GasId[];
  return produced.every((g) => allowedGases.includes(g));
}

export const CHALLENGES: Challenge[] = [
  {
    id: 'ice_age',
    name: 'Ice Age',
    description: 'Reach a strong industrial base while keeping temperature below +2°C.',
    icon: '🧊',
    restriction: { maxTemperatureC: 2 },
    goal: { description: 'Reach civilization level 40 without exceeding +2°C.', check: (_s, civLevel) => civLevel >= 40 },
    rewardDescription: '+10% permanent production',
    reward: { globalProductionMultiplier: 1.1 },
  },
  {
    id: 'low_carbon',
    name: 'Low Carbon',
    description: 'Coal technologies are forbidden for the entire run.',
    icon: '🚫',
    restriction: {
      disabledTechIds: ALL_TECHNOLOGIES.filter((t) => t.id.includes('coal')).map((t) => t.id),
    },
    goal: { description: 'Accumulate 1 trillion (1e12) total Energy produced.', check: (s) => s.resources.energy.gte('1e12') },
    rewardDescription: '+25% Research, permanently',
    reward: { researchMultiplier: 1.25 },
  },
  {
    id: 'no_oil',
    name: 'No Oil',
    description: 'Oil and its downstream technologies are off-limits.',
    icon: '🛢️',
    restriction: {
      disabledTechIds: ALL_TECHNOLOGIES.filter((t) => t.branch === 'fossilFuels' || ['automobile', 'diesel_engine', 'jet_aircraft', 'petrochemicals'].includes(t.id)).map((t) => t.id),
    },
    goal: { description: 'Reach civilization level 35 without any oil technology.', check: (_s, civLevel) => civLevel >= 35 },
    // The copy here advertised a Transportation branch bonus that the effect
    // below never granted. The effect is the shipped balance and is left as it
    // is; the text is what was wrong, so the text is what changed. Kept in
    // step with the Kotlin port, which pins this string in ContentParityTest.
    rewardDescription: '+5% permanent production',
    reward: { globalProductionMultiplier: 1.05 },
  },
  {
    id: 'primitive',
    name: 'Primitive',
    description: 'Never advance past the Iron Age.',
    icon: '⛏️',
    restriction: { maxTechTier: 6 },
    goal: { description: 'Reach 400 ppm atmospheric CO₂ using only Primitive-era technology.', check: (s) => s.atmosphere.co2.gte(120) },
    rewardDescription: 'Start every future run with extra starting Energy',
    reward: { startingResources: { energy: 500 } },
  },
  {
    id: 'speedrun',
    name: 'Speedrun',
    description: 'Reset Earth as quickly as possible.',
    icon: '⏱️',
    restriction: {},
    goal: { description: 'Reset within 15 minutes of starting the run.', check: (s) => (Date.now() - s.runStartedAt) / 1000 <= 900 },
    // A x1.0 multiplier is not a bonus, and this copy promised one anyway:
    // there is no 'Earth Points earned' field in the effect shape for it to
    // have been wired to. Described for what it is rather than inventing a
    // balance change. See docs/CODEBASE_AUDIT.md in the Android repository.
    rewardDescription: 'Bragging rights \u2014 this one pays no permanent bonus',
    reward: { globalProductionMultiplier: 1.0 },
  },
  {
    id: 'single_gas',
    name: 'Single Gas',
    description: 'Only technologies that emit CO₂ (or nothing) may be used.',
    icon: '☁️',
    restriction: {
      disabledTechIds: ALL_TECHNOLOGIES.filter((t) => !techProducesOnly(t, ['co2'])).map((t) => t.id),
    },
    goal: { description: 'Reach civilization level 30 using only CO₂-producing technology.', check: (_s, civLevel) => civLevel >= 30 },
    rewardDescription: '+30% CO₂ production, permanently',
    reward: { gasProductionMultiplier: { gas: 'co2', multiplier: 1.3 } },
  },
  {
    id: 'methane_world_challenge',
    name: 'Methane World',
    description: 'Make CH₄ the dominant greenhouse gas by the time you reset.',
    icon: '🐄',
    restriction: {},
    goal: {
      description: 'Reset while CH₄ contributes more forcing than any other gas.',
      check: (s) => s.forcing.perGas.ch4 > s.forcing.perGas.co2 && s.forcing.perGas.ch4 > 0,
    },
    rewardDescription: '+50% CH₄ production, permanently',
    reward: { gasProductionMultiplier: { gas: 'ch4', multiplier: 1.5 } },
  },
  {
    id: 'zero_emissions',
    name: 'Zero Emissions',
    description: 'Build a strong civilization while keeping gross emissions minimal.',
    icon: '🌱',
    restriction: { disabledTechIds: ['coal_power_plant', 'supercritical_coal', 'coal_mega_mining', 'oil_sands'] },
    goal: {
      description: 'Reach civilization level 25 while total gross gas production stays under 1e6 kg/s.',
      check: (s, civLevel) => civLevel >= 25,
    },
    rewardDescription: '+20% Research, permanently',
    reward: { researchMultiplier: 1.2 },
  },
];

export const CHALLENGE_BY_ID: Record<string, Challenge> = Object.fromEntries(CHALLENGES.map((c) => [c.id, c]));

/** Whether purchasing `techId` would violate the active challenge's restriction. */
export function isTechRestrictedByChallenge(challenge: Challenge, techId: string): boolean {
  const tech = TECH_BY_ID[techId];
  if (!tech) return false;
  if (challenge.restriction.disabledTechIds?.includes(techId)) return true;
  if (challenge.restriction.maxTechTier !== undefined && tech.tier > challenge.restriction.maxTechTier) return true;
  if (challenge.restriction.onlyGasIds && tech.effect.gasProductionPerUnit) {
    if (!techProducesOnly(tech, challenge.restriction.onlyGasIds)) return true;
  }
  return false;
}

export function disabledTechIdsForChallenge(challenge: Challenge | null): Set<string> {
  if (!challenge) return new Set();
  return new Set(ALL_TECHNOLOGIES.filter((t) => isTechRestrictedByChallenge(challenge, t.id)).map((t) => t.id));
}

/** Whether the challenge's live restriction (e.g. a temperature ceiling) has been broken this run. */
export function isChallengeRestrictionViolated(challenge: Challenge, state: GameState): boolean {
  if (challenge.restriction.maxTemperatureC !== undefined && state.temperatureAnomalyC > challenge.restriction.maxTemperatureC) {
    return true;
  }
  return false;
}

export function isChallengeGoalMet(challenge: Challenge, state: GameState, civLevel: number): boolean {
  return challenge.goal.check(state, civLevel);
}

/** Aggregates every completed challenge's permanent reward into one effect bundle, in the same shape prestige upgrades use. */
export function computeChallengeRewardEffects(completed: Record<string, boolean>): PrestigeUpgradeEffect[] {
  return CHALLENGES.filter((c) => completed[c.id]).map((c) => c.reward);
}
